generate_qfp.py

```diff
@@ -0,0 +1,571 @@
+    0.40: Excess(0.20, 0.30,  0.00, 0.40),
+    0.35: Excess(0.20, 0.30, -0.01, 0.40)
+}
+DENSITY_LEVEL_B = {  # Nominal
+    1.00: Excess(0.35, 0.45,  0.06, 0.20),
+    0.80: Excess(0.30, 0.40,  0.05, 0.20),
+    0.65: Excess(0.25, 0.35,  0.04, 0.20),
+    0.50: Excess(0.20, 0.30,  0.02, 0.20),
+    0.40: Excess(0.15, 0.25, -0.01, 0.20),
+    0.35: Excess(0.15, 0.25, -0.02, 0.20)
+}
+DENSITY_LEVEL_C = {  # Least
+    1.00: Excess(0.30, 0.40,  0.05, 0.10),
+    0.80: Excess(0.25, 0.35,  0.04, 0.10),
+    0.65: Excess(0.20, 0.30,  0.03, 0.10),
+    0.50: Excess(0.15, 0.25,  0.01, 0.10),
+    0.40: Excess(0.10, 0.20, -0.02, 0.10),
+    0.35: Excess(0.10, 0.20, -0.03, 0.10)
+}
+
+
+class QfpConfig:
+    def __init__(
+        self,
+        name: str,  # e.g. QFP or LQFP
+        body_size_x: float,
+        body_size_y: float,
+        height: float,  # Total height, A in the datasheet
+        pitch: float,
+        lead_count: int,
+        lead_span_x: float,  # Total length from tip to tip, D in the datasheet
+        lead_span_y: float,  # Total length from tip to tip, E in the datasheet
+        lead_width: float,  # b in the datasheet, nominal
+        keywords: str,
+        name_prefix: Optional[str] = None,  # E.g. the manufacturer name
+    ):
+        assert body_size_x == body_size_y, 'rectangular support not yet implemented'
+
+        self.name = name
+        self.body_size_x = body_size_x
+        self.body_size_y = body_size_y
+        self.pitch = pitch
+        self.lead_count = lead_count
+        self.height = height
+        self.lead_span_x = lead_span_x
+        self.lead_span_y = lead_span_y
+        self.lead_width = lead_width
+        self.keywords = keywords
+        self.name_prefix = name_prefix or ''
+
+        # Common dimensions
+        self.lead_contact_length = 0.60  # L in the datasheet
+
+        # Calculate lead length
+        self.lead_length_x = (lead_span_x - body_size_x) / 2
+        self.lead_length_y = (lead_span_y - body_size_y) / 2
+
+    def get_configs(self) -> List['QfpConfig']:
+        return [self]
+
+    def ipc_name(self):
+        return '{}{}{}P{}X{}X{}-{}'.format(
+            self.name_prefix,
+            self.name,
+            fd(self.pitch),
+            fd(self.lead_span_x),
+            fd(self.lead_span_y),
+            fd(self.height),
+            self.lead_count,
+        )
+
+    def description(self):
+        if self.name == 'LQFP':
+            full_name = 'Low Profile Quad Flat Package (LQFP)'
+        elif self.name == 'TQFP':
+            full_name = 'Thin Profile Quad Flat Package (TQFP)'
+        elif self.name == 'QFP':
+            full_name = 'Quad Flat Package (QFP)'
+        else:
+            raise ValueError('Invalid name: {}'.format(self.name))
+        return '{}-pin {}, standardized by JEDEC in MS-026.\\n\\n' \
+               'Pitch: {} mm\\nBody size: {}x{} mm\\nLead span: {}x{} mm\\n' \
+               'Max height: {} mm\\n\\nGenerated with {}'.format(
```

Comment:
Why "Max height"? I think the value is actually the nominal height (which is also included in the package's name), so I would write just "Height":

```suggestion
               'Height: {} mm\\n\\nGenerated with {}'.format(
```

Reply:
You're right, that's a copy pasta error.

Reply:
Ah, no, I checked again, and JEDEC only specifies max height, not nominal height:

![img](https://tmp.dbrgn.ch/screenshots/20190217234146-34ronglc.png)

(Row A)

In that case I think we should stick with "Max height" in the description, right?

Reply:
Yeah but IMHO we should specify nominal values instead of maximum values, otherwise it's confusing if some packages specify nominal values and some specify maximum values. When searching in the Internet LQFP seems to mean 1.4mm and TQFP 1.0mm, thus I would just take these two values. When considering +/-0.2mm tolerance, it also matches JEDEC's maximum values of 1.6mm and 1.2mm, which sounds reasonable to me ;)

Reply:
Ah, I didn't realize that I already used the nominal values in the generator, and not A from the datasheet :slightly_smiling_face: 

Ok fixed. I separately list nominal and max height now.